src/_P103_Atlas_EZO_pH_ORP_EC.ino

```diff
@@ -34,496 +36,528 @@ boolean Plugin_103(byte function, struct EventStruct *event, String& string)
+
+      if (isFormItemChecked(F("Plugin_103_enable_cal_dry")))
+      {
+        cmd += F("dry");
+        triggerCalibrate = true;
+      }
+      else if (isFormItemChecked(F("Plugin_103_enable_cal_single")))
+      {
+        cmd += PCONFIG(3);
+        triggerCalibrate = true;
+      }
+      else if (isFormItemChecked(F("Plugin_103_enable_cal_L")))
+      {
+        cmd += F("low,");
+        cmd += PCONFIG(4);
+        triggerCalibrate = true;
+      }
+      else if (isFormItemChecked(F("Plugin_103_enable_cal_H")))
+      {
+        cmd += F("high,");
+        cmd += PCONFIG(5);
+        triggerCalibrate = true;
+      }
       break;
     }
+    }
 
-    case PLUGIN_READ:
+    if (triggerCalibrate)
     {
-      board_type = PCONFIG(0);
-      I2Cchoice  = PCONFIG(1);
+      char calibration[ATLAS_EZO_RETURN_ARRAY_SIZE] = {0};
+      _P103_send_I2C_command(I2Cchoice, cmd, calibration);
+    }
 
-      char sensordata[32];
-      bool status;
-      String readCommand;
+    if (board_type == PH || board_type == EC)
+    {
+      char deviceTemperatureTemplate[40] = {0};
+      String tmpString = web_server.arg(F("Plugin_103_temperature_template"));
+      strncpy(deviceTemperatureTemplate, tmpString.c_str(), sizeof(deviceTemperatureTemplate) - 1);
+      ZERO_TERMINATE(deviceTemperatureTemplate); // be sure that our string ends with a \0
 
-      if(board_type == PH || board_type == EC) {
-        // first set the temperature of reading
-        char deviceTemperatureTemplate[40];
-        LoadCustomTaskSettings(event->TaskIndex, (byte *)&deviceTemperatureTemplate, sizeof(deviceTemperatureTemplate));
+      addHtmlError(SaveCustomTaskSettings(event->TaskIndex, (byte *)&deviceTemperatureTemplate, sizeof(deviceTemperatureTemplate)));
+    }
 
-        String deviceTemperatureTemplateString(deviceTemperatureTemplate);
-        String pooltempString(parseTemplate(deviceTemperatureTemplateString, 40));
+    success = true;
+    break;
+  }
 
-        readCommand = F("RT,");
-        double  temperatureReading;
+  case PLUGIN_INIT:
+  {
+    break;
+  }
 
-        if (Calculate(pooltempString.c_str(), temperatureReading) != CalculateReturnCode::OK) {
-            temperatureReading = FIXED_TEMP_VALUE;
-        }
+  case PLUGIN_READ:
+  {
+    board_type = PCONFIG(0);
+    I2Cchoice = PCONFIG(1);
 
-        readCommand += temperatureReading;
+    String readCommand;
 
-      } else if(board_type == ORP) {
-        readCommand = F("R,");
-      }
+    if (board_type == PH || board_type == EC)
+    {
+      // first set the temperature of reading
+      char deviceTemperatureTemplate[40];
+      LoadCustomTaskSettings(event->TaskIndex, (byte *)&deviceTemperatureTemplate, sizeof(deviceTemperatureTemplate));
 
-      // ok, now we can read the value
-      status = _P103_send_I2C_command(I2Cchoice, readCommand, sensordata);
+      String deviceTemperatureTemplateString(deviceTemperatureTemplate);
```

Comment:
No guarantee the char array is zero terminated.

Reply:
I will initlialize it with {0}

Reply:
You read all bytes from the settings, so better call ZERO_TERMINATE after loading it.

Reply:
OK

Reply:
fixed